File: Top Interview 150/[240722] 380. Insert Delete GetRandom O(1)/정훈.py
from random import randint

class RandomizedSet:
    rs: Set

    def __init__(self):
        self.rs = set()

    def insert(self, val: int) -> bool:
        if val not in self.rs:
            self.rs.add(val)
            return True
        return False

    def remove(self, val: int) -> bool:
        if val not in self.rs:
            return False
        self.rs.remove(val)
        return True

    def getRandom(self) -> int:
        ri = randint(0, len(self.rs)-1)
        return list(self.rs)[ri]

File: Top Interview 150/[240722] 380. Insert Delete GetRandom O(1)/정훈.java
class RandomizedSet {
    private ArrayList<Integer> list;
    private Map<Integer,Integer> map;

    public RandomizedSet() {
        list = new ArrayList<>();
        map = new HashMap<>();
    }
    
    public boolean insert(int val) {
        if (map.containsKey(val))
            return false;
        list.add(val);
        map.put(val,list.size() - 1);
        return true;
    }
    
    public boolean remove(int val) {
        if (!map.containsKey(val))
            return false;
        list.set(map.get(val), list.get(list.size()-1));
        map.put(list.get(map.get(val)), map.get(val));
        list.remove(list.size()-1);
        map.remove(val);
        return true;
    }
    
    public int getRandom() {
        Random rand = new Random();
        return list.get(rand.nextInt(list.size()));
    }
}
